sql: support SECURITY DEFINER for routines

annrpom opened the pull request:
### sql: correct error code

This patch assigns the pgcode (42501) to an ownership 
privilege check error. This aligns with postgres behavior.

```
postgres=> alter function testfn1(int) security definer;
ERROR:  42501: must be owner of function testfn1
LOCATION:  aclcheck_error, aclchk.c:3790
```

Release note: None

---

### opt/cat: add sentinel const for StableID

This patch adds the `DefaultStableID` constant
of 0 to describe an uninitialized stable ID for
catalog objects. GlobalPrivilege objects always
return this value as their ID.

Epic: none

Release note: None

---

### opt/cat: refactor priv checks to specify a user

This patch refactors privilege check methods of the `catalog`
interface to include a `username.SQLUsername`. This provides us
with the new API:

```
CheckPrivilege(ctx context.Context, o Object, user username.SQLUsername, priv privilege.Kind) error

CheckExecutionPrivilege(ctx context.Context, oid oid.Oid, user username.SQLUsername) error
```

Release note: None

---

### sql: support SECURITY DEFINER for routines

Epic: [CRDB-37477](https://cockroachlabs.atlassian.net/browse/CRDB-37477)
Fixes: https://github.com/cockroachdb/cockroach/issues/121514

Release note (sql change): Added support for `SECURITY DEFINER` in user defined
functions (UDFs) and stored procedures (SPs). When a UDF/SP (routine) is created
with `SECURITY DEFINER`, at execution, the privileges of the owner will be
checked.

Routines can now specify `[EXTERNAL] SECURITY INVOKER` (this is the
default -- privileges of the invoker are checked at execution) or
`[EXTERNAL] SECURITY DEFINER`. Note that the `EXTERNAL` keyword is
optional and solely exists for sql conformity.

In addition, altering a UDF's security "mode" is accomplished by:
`ALTER FUNCTION ... [EXTERNAL] SECURITY {INVOKER/DEFINER}`.

annrpom committed:
sql: correct error code

This change aligns with postgres behavior.
postgres:
```
postgres=> alter function testfn1(int) security definer;
ERROR:  42501: must be owner of function testfn1
LOCATION:  aclcheck_error, aclchk.c:3790
```

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/129720)
<!-- Reviewable:end -->

mw5h reviewed (dismissed):
> Reviewed 1 of 1 files at r1, 8 of 18 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129720)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom, @mgartner, and @rafiss)

___
*[`pkg/sql/opt_catalog.go` line 426 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/129720#-O5PlVYD7HEoFLvToPQX:-O5PlVYD7HEoFLvToPQY:bggsskz) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffb55aa1caa4e5e961e8e8b21a46d3f00ccdb380/pkg/sql/opt_catalog.go#L426)):*
> ```Go
> 	}
> 	creator := funcDesc.FuncDesc().Privileges.Owner()
> 	if o.ID() == 0 {
> ```

Could we use a more descriptive sentinel here than '0'?


<!-- Sent from Reviewable.io -->

annrpom committed:
opt/cat: add sentinel for StableID

This patch adds the `DefaultStableID` constant
of 0 to describe an uninitialized stable ID for
catalog objects. GlobalPrivilege objects always
return this value as their ID.

Epic: none

Release note: None

annrpom reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129720)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @mw5h, and @rafiss)

___
*[`pkg/sql/opt_catalog.go` line 426 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/129720#-O5PlVYD7HEoFLvToPQX:-O5PpKHn4JqLz2aK6nSd:b-qx102m) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffb55aa1caa4e5e961e8e8b21a46d3f00ccdb380/pkg/sql/opt_catalog.go#L426)):*
<details><summary><i>Previously, mw5h (Matt White) wrote…</i></summary><blockquote>

Could we use a more descriptive sentinel here than '0'?
</blockquote></details>

done -- i added `cat.InvalidStableID` and found 2 (?) other places where we can use the more descriptive sentinel instead


<!-- Sent from Reviewable.io -->

rafiss reviewed (changes requested):
nice work so far!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129720#-:-O5Pypf5ClqSKMkMOnef:b3b5okd)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom, @mgartner, and @mw5h)

___
*[`pkg/sql/opt_catalog.go` line 427 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/129720#-O5PvkjZ1RYym2vYdjkk:-O5Pvkj_DBgTKAmGTjat:b-rut6os) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bfe69dd880334ff6c0346cfe6967c5662774f68/pkg/sql/opt_catalog.go#L427)):*
> ```Go
> 	owner := funcDesc.FuncDesc().Privileges.Owner()
> 	if o.ID() == cat.InvalidStableID {
> 		return oc.planner.CheckPrivilegeForUser(ctx, syntheticprivilege.GlobalPrivilegeObject, priv, owner)
> ```

i didn't follow this part. which global privilege are we checking here? as i understand, we're trying to check the EXECUTE privilege on the function, but since EXECUTE is not a valid global privilege, this fallback doesn't feel correct.
___
*[`pkg/sql/opt_catalog.go` line 429 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/129720#-O5Py4a91TcqOfjk0uFf:-O5Py4a91TcqOfjk0uFg:bpgoipt) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bfe69dd880334ff6c0346cfe6967c5662774f68/pkg/sql/opt_catalog.go#L429)):*
> ```Go
> 		return oc.planner.CheckPrivilegeForUser(ctx, syntheticprivilege.GlobalPrivilegeObject, priv, owner)
> 	}
> 	desc, err := getDescFromCatalogObjectForPermissions(o)
> ```

do you know why this is needed? it looks like we already have the funcDesc above.
___
*[`pkg/sql/opt_catalog.go` line 446 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/129720#-O5PyOvg65CwG4cHsLOW:-O5PyOvg65CwG4cHsLOX:b-npzibx) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bfe69dd880334ff6c0346cfe6967c5662774f68/pkg/sql/opt_catalog.go#L446)):*
> ```Go
> 
> // CheckExecutionPrivilege is part of the cat.Catalog interface.
> func (oc *optCatalog) CheckExecutionPrivilege(ctx context.Context, oid oid.Oid) error {
> ```

it feels like this is the function that we should be modifying. the one above this looks like it's only used for tables.
___
*[`pkg/sql/opt/optbuilder/util.go` line 715 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/129720#-O5Pw_1e36t6rTjfhwwz:-O5Pw_1e36t6rTjfhwx-:b-wv8rz2) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bfe69dd880334ff6c0346cfe6967c5662774f68/pkg/sql/opt/optbuilder/util.go#L715)):*
> ```Go
> 		// If the routineOid is set, we look up the owner of the function and
> 		// run the privilege check against that user.
> 		if b.routineOid != nil {
> ```

i'll let the Queries team override me if they want, but it feels like this is a little hard to reason about. it's not intuitive that the presence/absence of `routineOid` is the thing that determines if we are checking the privileges for the invoker or definer. 

this might end up getting cleaned up anyway if we address the earlier comment i left on the `CheckExecutionPrivilege` function.

seems like it would be better to have an invoker/definer enum that we pass into `CheckExecutionPrivilege`.
___
*[`pkg/sql/sem/tree/overload.go` line 300 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/129720#-O5PwnC75R39QhYnRxAN:-O5PwnC75R39QhYnRxAO:b-e60a3o) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bfe69dd880334ff6c0346cfe6967c5662774f68/pkg/sql/sem/tree/overload.go#L300)):*
> ```Go
> 
> 	// UsesSecurityDefiner is true when privilege checks for function execution
> 	// should be performed against a specific role rather than the invoking user.
> ```

nit: would it be more clear to say "against the function owner" instead of "against a specific role"?
___
*[`pkg/sql/sem/tree/overload.go` line 301 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/129720#-O5PwcyL9NDmokZBenUW:-O5PwcyL9NDmokZBenUX:biu5875) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bfe69dd880334ff6c0346cfe6967c5662774f68/pkg/sql/sem/tree/overload.go#L301)):*
> ```Go
> 	// UsesSecurityDefiner is true when privilege checks for function execution
> 	// should be performed against a specific role rather than the invoking user.
> 	UsesSecurityDefiner bool
> ```

nit: do you know why this needs to be a bool, instead of using a `definer`/`invoker` enum?


<!-- Sent from Reviewable.io -->

annrpom reviewed (commented):
i left some comments unreplied to, but only due to the fact that they might change after my replies! i also think my wording on the description os what SECURITY DEFINER might need a rework?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129720#-:-O5Uuc9V8-8DDs4iJxu-:b-n2r9yy)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @mw5h, and @rafiss)

___
*[`pkg/sql/opt_catalog.go` line 427 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/129720#-O5PvkjZ1RYym2vYdjkk:-O5UovJX0tVmVBrL8B7P:bf4q4pd) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bfe69dd880334ff6c0346cfe6967c5662774f68/pkg/sql/opt_catalog.go#L427)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i didn't follow this part. which global privilege are we checking here? as i understand, we're trying to check the EXECUTE privilege on the function, but since EXECUTE is not a valid global privilege, this fallback doesn't feel correct.
</blockquote></details>

we're trying to check whichever privileges we need to be able to properly execute the body of the function: so if all the function does is select from a table, here, we check if the owner has privilege SELECT on the table descriptor

the EXECUTE privilege checking is handled elsewhere, but not apart of what makes a routine SECURITY DEFINER
___
*[`pkg/sql/opt_catalog.go` line 429 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/129720#-O5Py4a91TcqOfjk0uFf:-O5UsD_j5M-LbAG4BjOQ:b-7eeq0t) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bfe69dd880334ff6c0346cfe6967c5662774f68/pkg/sql/opt_catalog.go#L429)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

do you know why this is needed? it looks like we already have the funcDesc above.
</blockquote></details>

this is the descriptor of the actual object we are checking to see that we have a privilege on (let me know if i worded that weirdly). this object + the required privilege type is later on added to the builder's metadata here https://github.com/cockroachdb/cockroach/blob/c5fd50407101e63d8939fb5c1ec6161a9dc21d34/pkg/sql/opt/metadata.go#L139-L141
___
*[`pkg/sql/sem/tree/overload.go` line 300 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/129720#-O5PwnC75R39QhYnRxAN:-O5UuMdx4-4KRmVbYeoI:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bfe69dd880334ff6c0346cfe6967c5662774f68/pkg/sql/sem/tree/overload.go#L300)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: would it be more clear to say "against the function owner" instead of "against a specific role"?
</blockquote></details>

done
___
*[`pkg/sql/sem/tree/overload.go` line 301 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/129720#-O5PwcyL9NDmokZBenUW:-O5UvIvPDd37nUdurC6B:b-d2l2a7) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bfe69dd880334ff6c0346cfe6967c5662774f68/pkg/sql/sem/tree/overload.go#L301)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: do you know why this needs to be a bool, instead of using a `definer`/`invoker` enum?
</blockquote></details>

prototyping remnants perhaps -- let me fix


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129720)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom, @mgartner, and @mw5h)

___
*[`pkg/sql/opt_catalog.go` line 427 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/129720#-O5PvkjZ1RYym2vYdjkk:-O5UwKwJ0oOFLD0ajNGZ:b-xptid3) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bfe69dd880334ff6c0346cfe6967c5662774f68/pkg/sql/opt_catalog.go#L427)):*
<details><summary><i>Previously, annrpom (annie pompa) wrote…</i></summary><blockquote>

we're trying to check whichever privileges we need to be able to properly execute the body of the function: so if all the function does is select from a table, here, we check if the owner has privilege SELECT on the table descriptor

the EXECUTE privilege checking is handled elsewhere, but not apart of what makes a routine SECURITY DEFINER
</blockquote></details>

i see, i understand it better now. i still have the initial question though: why are we checking a global privilege here? in your example, we would want to check if the owner has the SELECT privilege on the table. so why are we falling back to checking a global privilege.

it feels like a better fallback is
```
if o.ID() == cat.InvalidStableID {
  return
}
```

perhaps one thing that could help is a comment that says in which cases we will see `o.ID() == cat.InvalidStableID`.


<!-- Sent from Reviewable.io -->

annrpom reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129720)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @mw5h, and @rafiss)

___
*[`pkg/sql/opt_catalog.go` line 427 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/129720#-O5PvkjZ1RYym2vYdjkk:-O5VK7TN3SrILqJNn6sd:b-gkabau) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bfe69dd880334ff6c0346cfe6967c5662774f68/pkg/sql/opt_catalog.go#L427)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i see, i understand it better now. i still have the initial question though: why are we checking a global privilege here? in your example, we would want to check if the owner has the SELECT privilege on the table. so why are we falling back to checking a global privilege.

it feels like a better fallback is
```
if o.ID() == cat.InvalidStableID {
  return
}
```

perhaps one thing that could help is a comment that says in which cases we will see `o.ID() == cat.InvalidStableID`.
</blockquote></details>

ah right: i was thinking about if we were to ever allow `GlobalPrivilege` stuff in routines like showing cluster settings:

```
CREATE FUNCTION show()  RETURNS VOID SECURITY DEFINER AS $$                               
    show all cluster settings;                                       
$$ LANGUAGE sql; 
```

iiuc, in this case, the o.ID() here would be our 0 value since we are dealing with a `GlobalPrivilege` object:
https://github.com/cockroachdb/cockroach/blob/4756b021edfdd477427f4ec34bf83f837f26a855/pkg/sql/syntheticprivilege/global_privilege.go#L67-L70

in other cases -- like when we are looking at tables, o.ID should not be our 0 value. then, we go ahead and get the descriptor from our catalog object `o` and then check the privilege (so from our continued example: SELECT) on the descriptor (table descriptor) for the owner

also: i do agree on the comment about the comment; let me add that in


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129720)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom, @mgartner, and @mw5h)

___
*[`pkg/sql/opt_catalog.go` line 427 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/129720#-O5PvkjZ1RYym2vYdjkk:-O5WTSWn8nc0mbmnFUp0:b-9b51rm) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bfe69dd880334ff6c0346cfe6967c5662774f68/pkg/sql/opt_catalog.go#L427)):*
<details><summary><i>Previously, annrpom (annie pompa) wrote…</i></summary><blockquote>

ah right: i was thinking about if we were to ever allow `GlobalPrivilege` stuff in routines like showing cluster settings:

```
CREATE FUNCTION show()  RETURNS VOID SECURITY DEFINER AS $$                               
    show all cluster settings;                                       
$$ LANGUAGE sql; 
```

iiuc, in this case, the o.ID() here would be our 0 value since we are dealing with a `GlobalPrivilege` object:
https://github.com/cockroachdb/cockroach/blob/4756b021edfdd477427f4ec34bf83f837f26a855/pkg/sql/syntheticprivilege/global_privilege.go#L67-L70

in other cases -- like when we are looking at tables, o.ID should not be our 0 value. then, we go ahead and get the descriptor from our catalog object `o` and then check the privilege (so from our continued example: SELECT) on the descriptor (table descriptor) for the owner

also: i do agree on the comment about the comment; let me add that in
</blockquote></details>

ah i see what you mean now. but we shouldn't check that here; all the system privilege checks are done during the sql execution phase, not here in the optimizer. for example: https://github.com/cockroachdb/cockroach/blob/271ede58f04fa4d48a85b03190a56cfca5b588e1/pkg/sql/show_cluster_setting.go#L182

the optimizer only does privilege checking for queries that interact with table data, i.e., DML queries (INSERT, SELECT, DELETE, UPDATE).

so i would be curious to know when we see `o.ID() == cat.InvalidStableID`. that will help us know if we should error here, or just return early.
`


<!-- Sent from Reviewable.io -->

mgartner reviewed (changes requested):
> Reviewed 1 of 1 files at r1, 18 of 18 files at r3, 3 of 3 files at r4, 8 of 18 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129720)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom, @mw5h, and @rafiss)

___
*[`pkg/sql/drop_function.go` line 194 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/129720#-O5xgQq4DucndjfWwg2Q:-O5xgQq5ENOoWfB8bglb:bc5bdb0) ([raw file](https://github.com/cockroachdb/cockroach/blob/c540810ee90ad02c17f6cc027e111638571e0b15/pkg/sql/drop_function.go#L194)):*
> ```Go
> 	}
> 	if !hasOwernship {
> 		return pgerror.Newf(pgcode.InsufficientPrivilege, "must be owner of function %s", fnDesc.GetName())
> ```

nit: Can you add a regression test for this change?
___
*[`pkg/sql/opt_catalog.go` line 426 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/129720#-O5xhjbc5lsKRGSBfQo0:-O5xhjbc5lsKRGSBfQo1:b-pqcv93) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bfe69dd880334ff6c0346cfe6967c5662774f68/pkg/sql/opt_catalog.go#L426)):*
> ```Go
> 	}
> 	owner := funcDesc.FuncDesc().Privileges.Owner()
> 	if o.ID() == cat.InvalidStableID {
> ```

I'm not convinced `InvalidStableID` is the best name for this. If we have legitimate cases where `o.ID()` is `0`, then it is "valid". 
___
*[`pkg/sql/opt/optbuilder/util.go` line 715 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/129720#-O5Pw_1e36t6rTjfhwwz:-O5xiNn41rnJl_edP1r_:bpkyjdf) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bfe69dd880334ff6c0346cfe6967c5662774f68/pkg/sql/opt/optbuilder/util.go#L715)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i'll let the Queries team override me if they want, but it feels like this is a little hard to reason about. it's not intuitive that the presence/absence of `routineOid` is the thing that determines if we are checking the privileges for the invoker or definer. 

this might end up getting cleaned up anyway if we address the earlier comment i left on the `CheckExecutionPrivilege` function.

seems like it would be better to have an invoker/definer enum that we pass into `CheckExecutionPrivilege`.
</blockquote></details>

I'd prefer to use `b.insideUDF` and `b.insideSQLRoutine` to determine if we are building expressions inside of a routine definition, and replace `b.routineOid` with `b.privilegeOverride` that is always set when building a routine, either to the function invoker or the function definer.

Or alternatively, simply have `b.checkPrivilegeUser` that is _always_ set (even outside of UDFs) and always passed to `CheckPrivilege`, and modified to the definer or invoker when building UDFs. Then you can get rid of `CheckPrivilegeForRoutineOwner` and add a method that simply returns the privilege user to check for a routine based on the `SECURITY` setting.
___
*[`pkg/sql/logictest/testdata/logic_test/udf_security` line 149 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/129720#-O5xk7im3zpHa7P4r2E-:-O5xk7im3zpHa7P4r2E0:ba8e77w) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bfe69dd880334ff6c0346cfe6967c5662774f68/pkg/sql/logictest/testdata/logic_test/udf_security#L149)):*
> ```
> $$;
> 
> subtest end
> ```

I think we should have some tests with nested UDFs with various combinations of INVOKER/DEFINER and privileges to tables and the UDFs.


<!-- Sent from Reviewable.io -->

annrpom committed:
sql: support SECURITY DEFINER for routines

Epic: CRDB-37477
Fixes: #121514

Release note (sql change): Added support for `SECURITY DEFINER` in user defined
functions (UDFs) and stored procedures (SPs). When a UDF/SP (routine) is created
with `SECURITY DEFINER`, at execution, the privileges of the owner will be
checked.

Routines can now specify `[EXTERNAL] SECURITY INVOKER` (this is the
default -- privileges of the invoker are checked at execution) or
`[EXTERNAL] SECURITY DEFINER`. Note that the `EXTERNAL` keyword is
optional and solely exists for sql conformity.

In addition, altering a UDF's security "mode" is accomplished by:
`ALTER FUNCTION ... [EXTERNAL] SECURITY {INVOKER/DEFINER}`.

annrpom reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129720)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @mw5h, and @rafiss)

___
*[`pkg/sql/drop_function.go` line 194 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/129720#-O5xgQq4DucndjfWwg2Q:-O6X8NDZ3C9kQxSDZE8f:b-xze0le) ([raw file](https://github.com/cockroachdb/cockroach/blob/c540810ee90ad02c17f6cc027e111638571e0b15/pkg/sql/drop_function.go#L194)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: Can you add a regression test for this change?
</blockquote></details>

done -- i moved the regression test from `udf_security` to `drop_function` since it makes more sense to test in the latter
___
*[`pkg/sql/opt_catalog.go` line 426 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/129720#-O5xhjbc5lsKRGSBfQo0:-O6bdmdY7Acej1kHFzLi:bw91w3l) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bfe69dd880334ff6c0346cfe6967c5662774f68/pkg/sql/opt_catalog.go#L426)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I'm not convinced `InvalidStableID` is the best name for this. If we have legitimate cases where `o.ID()` is `0`, then it is "valid".
</blockquote></details>

fair point-- maybe a more accurate name is `DefaultStableID`; i'll name it that for now & am open to any objections
___
*[`pkg/sql/opt_catalog.go` line 427 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/129720#-O5PvkjZ1RYym2vYdjkk:-O6XY1oV5-g4gqkqS8Ty:b-qky3z8) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bfe69dd880334ff6c0346cfe6967c5662774f68/pkg/sql/opt_catalog.go#L427)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

ah i see what you mean now. but we shouldn't check that here; all the system privilege checks are done during the sql execution phase, not here in the optimizer. for example: https://github.com/cockroachdb/cockroach/blob/271ede58f04fa4d48a85b03190a56cfca5b588e1/pkg/sql/show_cluster_setting.go#L182

the optimizer only does privilege checking for queries that interact with table data, i.e., DML queries (INSERT, SELECT, DELETE, UPDATE).

so i would be curious to know when we see `o.ID() == cat.InvalidStableID`. that will help us know if we should error here, or just return early.
`
</blockquote></details>

We shouldn't see `o.ID() == cat.InvalidStableID` in this function then; it looks like there are only two times where we should expect to see o.ID be equal to `0`. Those are:
1. for the global default case
2. if we are attempting to decode a plan gist for a plan with a dropped table/dropped table indexes

I'll add this as a programming error
___
*[`pkg/sql/opt/optbuilder/util.go` line 715 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/129720#-O5Pw_1e36t6rTjfhwwz:-O6YOSuH2NKrSYc4ZDpx:bajvr90) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bfe69dd880334ff6c0346cfe6967c5662774f68/pkg/sql/opt/optbuilder/util.go#L715)):*
thank you both; thank you marcus for your detailed suggestions! 

> I'd prefer to use `b.insideUDF` and `b.insideSQLRoutine` to determine if we are building expressions inside of a routine definition, and replace `b.routineOid` with `b.privilegeOverride` that is always set when building a routine, either to the function invoker or the function definer.

what do we think about this current approach (adding some "PrivilegeOverride" struct that keeps track of the `tree.RoutineSecurity`, along with a `routineOid`)? we'll need this oid to be able [to get the function descriptor](https://github.com/annrpom/cockroach/blob/2bfe69dd880334ff6c0346cfe6967c5662774f68/pkg/sql/opt_catalog.go#L421) -- which has the owner of the function

for:

> Or alternatively, simply have `b.checkPrivilegeUser` that is _always_ set (even outside of UDFs) and always passed to `CheckPrivilege`, and modified to the definer or invoker when building UDFs. Then you can get rid of `CheckPrivilegeForRoutineOwner` and add a method that simply returns the privilege user to check for a routine based on the `SECURITY` setting.

i see, but i'm not sure how to get the `b.checkPrivilegeUser` populated before we have a planner available to us; to my understanding, this would the current user by default https://github.com/cockroachdb/cockroach/blob/3d71cf3ee690d9a2f930ae8f611700126ff63d52/pkg/sql/planner.go#L657-L659

and in the case where we are `b.insideSqlRoutine` and our outerfunc is `SECURITY DEFINER`, this would be `f`'s (type catalog.FunctionDescriptor) `f.GetPrivileges().Owner()` -- which we would want to set right here
https://github.com/annrpom/cockroach/blob/2bfe69dd880334ff6c0346cfe6967c5662774f68/pkg/sql/opt/optbuilder/routine.go#L351-L364
___
*[`pkg/sql/logictest/testdata/logic_test/udf_security` line 149 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/129720#-O5xk7im3zpHa7P4r2E-:-O6bZYrj6eHhc6nECi6_:b-d42tc6) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bfe69dd880334ff6c0346cfe6967c5662774f68/pkg/sql/logictest/testdata/logic_test/udf_security#L149)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I think we should have some tests with nested UDFs with various combinations of INVOKER/DEFINER and privileges to tables and the UDFs.
</blockquote></details>

done & this was a good thing to test -- it made me realize we also need similar logic for `CheckExecutionPrivilege` for nested funcs


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> Reviewed 1 of 18 files at r5, 14 of 19 files at r6, 4 of 5 files at r7, 2 of 21 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129720)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom, @mw5h, and @rafiss)

___
*[`pkg/sql/opt_catalog.go` line 426 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/129720#-O5xhjbc5lsKRGSBfQo0:-O6wuwqR5WA-8tLe05Jj:b-81nlyp) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bfe69dd880334ff6c0346cfe6967c5662774f68/pkg/sql/opt_catalog.go#L426)):*
<details><summary><i>Previously, annrpom (annie pompa) wrote…</i></summary><blockquote>

fair point-- maybe a more accurate name is `DefaultStableID`; i'll name it that for now & am open to any objections
</blockquote></details>

SGTM!
___
*[`pkg/sql/opt/metadata.go` line 512 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/129720#-O6wzKEI1FQAVCTQYQCQ:-O6wzKEJ0wvlE6NujeIj:bl97sh4) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5b25fdf81a8f220bfbe2494d0ef021589f98774/pkg/sql/opt/metadata.go#L512)):*
> ```Go
> 	}
> 	for _, dep := range md.udfDeps {
> 		if err := optCatalog.CheckExecutionPrivilege(ctx, dep.overload.Oid, nil); err != nil {
> ```

Oof, this is a tricky one. In this case we're attempting to reuse a previously built query plan. We need to make sure that the user (either the definer or invoker, depending on the function option) still has the correct privileges to invoke the function. If they don't, then the cached query plan cannot be reused because it would incorrectly give them access to the function. 

So I believe we need some way to keep track of which user was originally checked when the query plan was first built with optbuilder.

Also, I think we need something similar for data source privileges above—we can't just check the current user, it may be the definer user that originally had access to the table referenced inside the UDF. 

I'll think of some ideas on how to address this and maybe we can discuss in the next day or so.
___
*[`pkg/sql/opt/optbuilder/routine.go` line 62 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/129720#-O6wySMV3PDIjyJJzjAJ:-O6wySMV3PDIjyJJzjAK:b-xcyffz) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5b25fdf81a8f220bfbe2494d0ef021589f98774/pkg/sql/opt/optbuilder/routine.go#L62)):*
> ```Go
> 		if b.insideSQLRoutine {
> 			privilegeOverride = &b.privilegeOverride
> 		}
> ```

Re: my other suggestion to always have a "current user" stored in the `Builder` struct and having the catalog functions always take the user as input is that this gets clean up a bit to:

```go
if err := b.catalog.CheckExecutionPrivilege(b.ctx, o.Oid, b.currentUser); err != nil { /* ... */ }
```
___
*[`pkg/sql/opt/optbuilder/util.go` line 715 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/129720#-O5Pw_1e36t6rTjfhwwz:-O6wwsB9CRXK-LqYcBQp:b-g33zs1) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bfe69dd880334ff6c0346cfe6967c5662774f68/pkg/sql/opt/optbuilder/util.go#L715)):*
> i see, but i'm not sure how to get the `b.checkPrivilegeUser` populated before we have a planner available to us; to my understanding, this would the current user by default

You'd need to add a method like `GetCurrentUser` to the `cat.Catalog` interface, which the `optCatalog` should be able to implement since it has access to the planner. 

> what do we think about this current approach (adding some "PrivilegeOverride" struct that keeps track of the `tree.RoutineSecurity`, along with a `routineOid`)? we'll need this oid to be able [to get the function descriptor](https://github.com/annrpom/cockroach/blob/2bfe69dd880334ff6c0346cfe6967c5662774f68/pkg/sql/opt_catalog.go#L421) -- which has the owner of the function

I don't see the need to keep track of the `tree.RoutineSecurity` and pass along to the catalog methods. I think adjust the `CheckPrivilege...` methods to always take a user is sufficient and makes for a simpler API: 

```go
CheckPrivilege(ctx context.Context, o Object, user username.SQLUsername, priv privilege.Kind) error

CheckAnyPrivilege(ctx context.Context, o Object, user username.SQLUsername) error

CheckExecutionPrivilege(ctx context.Context, oid oid.Oid, user username.SQLUsername) error
```

(Ideally we could get rid of `CheckExecutionPrivilege` and just use `CheckPrivilege`, but it looks like there's a few hurdles to overcome before we can do that.)

What do you think of this approach?
___
*[`pkg/sql/opt/optbuilder/util.go` line 717 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/129720#-O6wwSHx8idF4PabMhS3:-O6wwSHx8idF4PabMhS4:b-fpcfb3) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5b25fdf81a8f220bfbe2494d0ef021589f98774/pkg/sql/opt/optbuilder/util.go#L717)):*
> ```Go
> 		// privilegeOverride's securityMode is RoutineDefiner, we will run the
> 		// privilege check against the owner of the routine.
> 		if b.insideSQLRoutine && b.privilegeOverride.SecurityMode == tree.RoutineDefiner {
> ```

I think you just want to check `b.insideUDF` which is incorrectly named, not `b.insideSQLRoutine`, but is true when building inside a UDF or stored procedure. `b.insideSQLRoutine` is not sufficient because it is `false` when building a PL/pgSQL UDF or SP. Sorry for leading you in the wrong direction there.

I'll put up a separate PR to fix the naming of `b.insideUDF`.
___
*[`pkg/sql/logictest/testdata/logic_test/udf_security` line 1 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/129720#-O6wuQqb2g8XKwiN42W2:-O6wuQqb2g8XKwiN42W3:b-8ek19m) ([raw file](https://github.com/cockroachdb/cockroach/blob/16c7950ddc1644b857298ca46d121c98d8585206/pkg/sql/logictest/testdata/logic_test/udf_security#L1)):*
> ```
> # LogicTest: !local-mixed-24.1 !local-mixed-24.2
> ```

nit: I think this test file was accidentally included in the first commit


<!-- Sent from Reviewable.io -->

annrpom committed:
opt/cat: refactor priv checks to specify a user

This patch refactors privilege check methods of the `catalog`
interface to include a `username.SQLUsername`. This provides us
with the new API:

```
CheckPrivilege(ctx context.Context, o Object, user username.SQLUsername, priv privilege.Kind) error

CheckExecutionPrivilege(ctx context.Context, oid oid.Oid, user username.SQLUsername) error
```

Release note: None

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129720)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom, @mw5h, and @rafiss)

___
*[`pkg/sql/opt/metadata.go` line 512 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/129720#-O6wzKEI1FQAVCTQYQCQ:-O83_HSt5nXVK-0SsmlU:bjk6s1s) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5b25fdf81a8f220bfbe2494d0ef021589f98774/pkg/sql/opt/metadata.go#L512)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Oof, this is a tricky one. In this case we're attempting to reuse a previously built query plan. We need to make sure that the user (either the definer or invoker, depending on the function option) still has the correct privileges to invoke the function. If they don't, then the cached query plan cannot be reused because it would incorrectly give them access to the function. 

So I believe we need some way to keep track of which user was originally checked when the query plan was first built with optbuilder.

Also, I think we need something similar for data source privileges above—we can't just check the current user, it may be the definer user that originally had access to the table referenced inside the UDF. 

I'll think of some ideas on how to address this and maybe we can discuss in the next day or so.
</blockquote></details>

We chatted IRL - my suggestions above were based on the incorrect assumption that changing privileges for a datasource would NOT update the datasource descriptor's version. I think privileges should work fine without doing anything special, given that the version IS updated, as long as we track all the datasource dependencies in `Metadata.dataSourceDeps`, which we already do AFAIK.


<!-- Sent from Reviewable.io -->

annrpom reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129720)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @mw5h, and @rafiss)

___
*[`pkg/sql/opt/optbuilder/util.go` line 715 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/129720#-O5Pw_1e36t6rTjfhwwz:-O8Cp3bP7zTZ5JfQWAgp:b-7n89tg) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bfe69dd880334ff6c0346cfe6967c5662774f68/pkg/sql/opt/optbuilder/util.go#L715)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

> i see, but i'm not sure how to get the `b.checkPrivilegeUser` populated before we have a planner available to us; to my understanding, this would the current user by default

You'd need to add a method like `GetCurrentUser` to the `cat.Catalog` interface, which the `optCatalog` should be able to implement since it has access to the planner. 

> what do we think about this current approach (adding some "PrivilegeOverride" struct that keeps track of the `tree.RoutineSecurity`, along with a `routineOid`)? we'll need this oid to be able [to get the function descriptor](https://github.com/annrpom/cockroach/blob/2bfe69dd880334ff6c0346cfe6967c5662774f68/pkg/sql/opt_catalog.go#L421) -- which has the owner of the function

I don't see the need to keep track of the `tree.RoutineSecurity` and pass along to the catalog methods. I think adjust the `CheckPrivilege...` methods to always take a user is sufficient and makes for a simpler API: 

```go
CheckPrivilege(ctx context.Context, o Object, user username.SQLUsername, priv privilege.Kind) error

CheckAnyPrivilege(ctx context.Context, o Object, user username.SQLUsername) error

CheckExecutionPrivilege(ctx context.Context, oid oid.Oid, user username.SQLUsername) error
```

(Ideally we could get rid of `CheckExecutionPrivilege` and just use `CheckPrivilege`, but it looks like there's a few hurdles to overcome before we can do that.)

What do you think of this approach?
</blockquote></details>

Ah right I forgot I can do that; what do you think about this change? We don't need `CheckAnyPrivilege` to change right now for security definer, but I'm open to changing it for future-proofing

I added a `checkPrivilegeUser` to the builder that will default to the current user upon init of the builder. I also added a `GetRoutineOwner` -- that helps us override the `checkPrivilegeUser` inside of `buildRoutine` if our routine's `SecurityMode` is security definer

This `checkPrivilegeUser` gets passed in to CheckPrivilege and CheckExecutionPrivilege
___
*[`pkg/sql/logictest/testdata/logic_test/udf_security` line 1 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/129720#-O6wuQqb2g8XKwiN42W2:-O8CrJ-m0pqb0BxFyN8k:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/16c7950ddc1644b857298ca46d121c98d8585206/pkg/sql/logictest/testdata/logic_test/udf_security#L1)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: I think this test file was accidentally included in the first commit
</blockquote></details>

fixed


<!-- Sent from Reviewable.io -->

annrpom commented:
@mgartner this is RFAL!

mgartner reviewed (changes requested):
Nicely done! Pulling out the refactoring into the first 3 commits made for a very pleasant review experience!

I believe I found 1 bug, see my comment below.

> Reviewed 25 of 25 files at r9, 6 of 6 files at r10, 12 of 12 files at r11, 19 of 20 files at r12, 1 of 1 files at r13, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129720#-:-O8dAVKdAboFj_c_nSRp:b-v2bgt8)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom, @mw5h, and @rafiss)

___
*[`pkg/sql/opt/optbuilder/routine.go` line 231 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/129720#-O8d6xf270zb0dQxK7pz:-O8d6xf270zb0dQxK7q-:b5lfwu2) ([raw file](https://github.com/cockroachdb/cockroach/blob/04000199b659ace65c592d0ebacbd7cfb64986f1/pkg/sql/opt/optbuilder/routine.go#L231)):*
> ```Go
> 			panic(err)
> 		}
> 		b.checkPrivilegeUser = checkPrivUser
> ```

We need to reset the user back to what it was previously after building the body of the function. I think we should move this block of code down to where other fields are reset in a `defer`: https://github.com/cockroachdb/cockroach/blob/9f61a38b670e57873509568f67bb665e50cd1935/pkg/sql/opt/optbuilder/routine.go#L341-L356

Can you also add this test case which shows the problem (the last SELECT should not be an error):

```sql
CREATE USER a;

CREATE USER b;

CREATE TABLE t (i INT);

GRANT SELECT ON t TO a;

SET ROLE b;

CREATE FUNCTION f() RETURNS INT LANGUAGE SQL SECURITY DEFINER AS $$
  SELECT 1
$$;

SET ROLE a;

SELECT i FROM t;
--   i
-- -----
-- (0 rows)

SELECT f()
UNION
SELECT i FROM t;
-- ERROR: user b does not have SELECT privilege on relation t
-- SQLSTATE: 42501
```


<!-- Sent from Reviewable.io -->

mgartner commented:
*[`pkg/sql/opt/optbuilder/routine.go` line 231 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/129720#-O8d6xf270zb0dQxK7pz:-O8dAvwiCerUj1rxTybK:b-oozb2x) ([raw file](https://github.com/cockroachdb/cockroach/blob/04000199b659ace65c592d0ebacbd7cfb64986f1/pkg/sql/opt/optbuilder/routine.go#L231)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

We need to reset the user back to what it was previously after building the body of the function. I think we should move this block of code down to where other fields are reset in a `defer`: https://github.com/cockroachdb/cockroach/blob/9f61a38b670e57873509568f67bb665e50cd1935/pkg/sql/opt/optbuilder/routine.go#L341-L356

Can you also add this test case which shows the problem (the last SELECT should not be an error):

```sql
CREATE USER a;

CREATE USER b;

CREATE TABLE t (i INT);

GRANT SELECT ON t TO a;

SET ROLE b;

CREATE FUNCTION f() RETURNS INT LANGUAGE SQL SECURITY DEFINER AS $$
  SELECT 1
$$;

SET ROLE a;

SELECT i FROM t;
--   i
-- -----
-- (0 rows)

SELECT f()
UNION
SELECT i FROM t;
-- ERROR: user b does not have SELECT privilege on relation t
-- SQLSTATE: 42501
```
</blockquote></details>

(Feel free to adjust the test case to use the `invoker`, `owner` pattern already used in your tests, if you'd like.)


<!-- Sent from Reviewable.io -->

annrpom reviewed (commented):
oh nice catch! sharp eyes 👀

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129720#-:-O8dC5Em4l2z2gHpXgNw:b2venea)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @mw5h, and @rafiss)

___
*[`pkg/sql/opt/optbuilder/routine.go` line 231 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/129720#-O8d6xf270zb0dQxK7pz:-O8dVnRjCAmvsKS_ydlb:b83pzye) ([raw file](https://github.com/cockroachdb/cockroach/blob/04000199b659ace65c592d0ebacbd7cfb64986f1/pkg/sql/opt/optbuilder/routine.go#L231)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

(Feel free to adjust the test case to use the `invoker`, `owner` pattern already used in your tests, if you'd like.)
</blockquote></details>

ah! good point; i think i get the gist of that test -- here is the test i added:

without the fix, if we have:
```
# assume we have a user 'owner' that has 2 functions:
# - f1() selects from table t and is SECURITY DEFINER
# - f2() selects from table t and is SECURITY INVOKER

SET ROLE invoker;

# there is no issue here even though there should be. we are accidentally escalating permissions for f2()
SELECT f1()
UNION
SELECT f2();
----
foo
```

with the fix:
```
# assume we have a user 'owner' that has 2 functions:
# - f1() selects from table t and is SECURITY DEFINER
# - f2() selects from table t and is SECURITY INVOKER

SET ROLE invoker;

statement error pgcode 42501 user invoker does not have SELECT privilege on relation t
SELECT f1()
UNION
SELECT f2();
----

```


<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice work!!!

> Reviewed 31 of 31 files at r14, 5 of 5 files at r15, 12 of 12 files at r16, 20 of 20 files at r17, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129720#-:-O8gqHkiEaGvwC-9ZsYH:b-3495r8)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @annrpom, @mw5h, and @rafiss)

___
*[`pkg/sql/opt/optbuilder/routine.go` line 231 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/129720#-O8d6xf270zb0dQxK7pz:-O8gpe3I5mnn1q-lh5C-:b-9fqeo9) ([raw file](https://github.com/cockroachdb/cockroach/blob/04000199b659ace65c592d0ebacbd7cfb64986f1/pkg/sql/opt/optbuilder/routine.go#L231)):*
<details><summary><i>Previously, annrpom (annie pompa) wrote…</i></summary><blockquote>

ah! good point; i think i get the gist of that test -- here is the test i added:

without the fix, if we have:
```
# assume we have a user 'owner' that has 2 functions:
# - f1() selects from table t and is SECURITY DEFINER
# - f2() selects from table t and is SECURITY INVOKER

SET ROLE invoker;

# there is no issue here even though there should be. we are accidentally escalating permissions for f2()
SELECT f1()
UNION
SELECT f2();
----
foo
```

with the fix:
```
# assume we have a user 'owner' that has 2 functions:
# - f1() selects from table t and is SECURITY DEFINER
# - f2() selects from table t and is SECURITY INVOKER

SET ROLE invoker;

statement error pgcode 42501 user invoker does not have SELECT privilege on relation t
SELECT f1()
UNION
SELECT f2();
----

```
</blockquote></details>

Great test!


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
nice work!

> Reviewed 1 of 25 files at r9, 1 of 31 files at r14, 3 of 5 files at r15, 7 of 12 files at r16, 20 of 20 files at r17, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129720#-:-O8grpbY-88n0w8j-ZLJ:b-j4xmku)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @annrpom and @mw5h)



<!-- Sent from Reviewable.io -->

annrpom commented:
TFTR! ('-')7 

bors r=mgartner, rafiss
